blog/content/post/2021-10-28-secure-jtag-on-imx-soc.md

```diff
@@ -0,0 +1,270 @@
+solution on i.MX6 and i.MX8MM. Probably this same feature is available on i.MX8
+but unfortunately security reference manual for this SoC are not publicly
+available.
+
+## Requirements
+
+- **Development board**: Any board or devkit with i.MX6 and i.MX8MM SoC. For
+testing this solution or i.MX6 we use our internal, custom board. For Secure
+JTAG on i.MX8M we use `DART-MX8M-MINI` from
+[Variscite](https://variwiki.com/index.php?title=DART-MX8M-MINI),
+
+- **SoC reference manual**: Setting JTAG in secure mode is a dangerous operation
+please check twice that everything is correct. Read SoC documentation and
+make sure that you will not brick the device,
+
+- **crucible tool**: JTAG mode and access key are saved in fuse bits - we need
+software to manipulate this in userspace. I used
+[crucible](https://github.com/f-secure-foundry/crucible/) for that, but I show
+you how to change one-time-programmable fields from U-boot,
+
+- **JTAG compatible with OpenOCD**: On IMX sites we can read that Secure JTAG is
+supported only by the Lauterbach environment and ARM-DS5 IDE with DSTREAM
+debugger. In 3mdeb we are always trying to use open-source software.
+Unfortunately, OpenOCD doesn't officially support Secure JTAG in i.MX SoC's,
+but I found not accepted yet patch from 2014 which add support for work with SJC
+(System JTAG controller) in a secure mode. This patch work on OpenOCD from
+[official repository](git://git.code.sf.net/p/openocd/code) if you checkout this
+with `970a12aef` hash. Is a chance to apply this in the actual version of
+OpenOCD, but it needs some code changes. As a hardware, i used `ARM-USB-OCD-H`
+from Olimex,
+
+- **System image**: Any Linux distribution with kernel module `nvmem-imx-ocotp`
+- for modify OTP memory.
+
+## How it works
+
+i.MX6 and i.MX8M offers JTAG in three modes:
+
+- **No debug** - security on the highest level, but after setting that
+development and debugging by JTAG port will be unavailable permanently on this
+device,
+
+- **Secure JTAG** - mode described in this article, the best compromise between
+security and development possibilities - only person with access key can use
+JTAG,
+
+- **Enabled JTAG** - default mode, enabled for anyone who has physical access to
+JTAG pins or even only device (is a chance to connect to the JTAG pins if the
+connector isn't soldered).
+
+Secure JTAG mode based on challenge/response mechanism. SoC has saved during
+manufacturing a unique challenge key. User can generate their response 56-bit
+key and program it in JTAG fuse `SJC_RESP`. While trying to access the JTAG
+port, SJC gives a challenge key. We can simplify this to the JTAG controller ID.
+Now user should pass the response key. It is compared with the response stored
+in SoC fuse bits. If keys are the same, JTAG is enabled. Bellow, you can see
+schemat described this mechanism:
+
+![Secure JTAG - how it works](/img/secure-jtag.png)
+
+JTAG mode can be set in `JTAG_SMODE` fuse by values:
```

Comment:
@kotylamichal maybe add some links to the documentation?

Reply:
ok